### PR TITLE
GSP-90: Re-support Initialization Via Connection String

### PR DESCRIPTION
**Pull request by xxchan**
related issue: https://github.com/beyondstorage/specs/issues/56

### CODE REVIEW

**Comment by xxchan** [comment]
demo: 
https://github.com/beyondstorage/go-storage/pull/589 
https://github.com/beyondstorage/go-service-qingstor/pull/54

**Comment by xxchan** [comment]
Oh, I found I missed the fact that *internally* service pairs are named with prefix `ty_` 😕 
 
image

So what should we expose to users?...

File: rfcs/90-parse-pairs-from-string.md [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+So I propose the following mechanism for parsing pairs:
+
+### Pair Registry
+
+We first register the types of global pairs and service pairs.
+
+From [go-storage] side:
+
+```go
+type PairMap map[string]reflect.Type
```
Comment (Xuanwo) [inline rfcs/90-parse-pairs-from-string.md:35]: Why we need to use `reflect.Type` here? The simple string maybe look enough to us? We only need to support limited types as described in specs.

File: rfcs/90-parse-pairs-from-string.md [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+```
+
+The pair information maps are also generated.
+
+### Parse According to Type
+
+For each type, we can have a parse function.
+
+```go 
+type parseFunc func(string) (interface{}, error)
```
Comment (Xuanwo) [inline rfcs/90-parse-pairs-from-string.md:58]: `parseFunc func(string) (interface{}, error)` not looks good to me.

`definitions` already has all information to parse the string into strongly typed pairs.

File: rfcs/90-parse-pairs-from-string.md [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+Then we can parse a pair of string into a pair. We check the pair's name in a `PairMap` and get its type information, then use corresponding parse function to parse the pair's value.
+
+```go
+func (m PairMap) parse(k string, v string) (pair Pair, err error) {}
+```
+
+Then we can easily provide APIs for users to parse pairs:
+
+```go 
+func Parse(ty string, k string, v string) (pair Pair, err error) {}
```
Comment (Xuanwo) [inline rfcs/90-parse-pairs-from-string.md:72]: I think the API exposed to users is `pairs.FromString(typ string, v string) (ps []Pair, err error)`.

And we don't need to support parse single pair: we only allow users to use these functions to do initialization.

File: rfcs/90-parse-pairs-from-string.md [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+1. Usability: If an application uses [go-storage]'s config string, then it either exposes the format of config string to end user directly (additional burden for end users) or it has to manipulate its config into our format (additional burden for application developers). And if the application wants to combine its other config into a config string, this is also inconvenient.
+2. Implementation: We didn't find a good way to parse pairs from strings easily and safely.
+
+So we let the users of [go-storage] construct pairs directly using `WithXxx()`. 
+
+But passing string config is indeed more convenient, if not much. And actually parsing string into pairs is not that difficult. We can use a registry mechanism like [GSP-48]'s implementation.
+
+## Proposal
+
+So I propose the following mechanism for parsing pairs:
```
Comment (Xuanwo) [inline rfcs/90-parse-pairs-from-string.md:26]: We need to define the format of our init string here.

File: rfcs/90-parse-pairs-from-string.md [hunk 5]
```diff
@@ -0,0 +1,106 @@
+---
+author: xxchan <xxchan22f@gmail.com>
+status: draft
+updated_at: 2021-06-01
+deprecates:
+  - 13-remove-config-string.md
+---
+
+# GSP-90: Parse Pairs from String
```
Comment (Xuanwo) [inline rfcs/90-parse-pairs-from-string.md:9]: This GSP intends to replace config string, it's not a general parser that parses the string into pairs.

File: rfcs/90-parse-pairs-from-string.md (hunk 2 above)
Comment (xxchan) [reply rfcs/90-parse-pairs-from-string.md:58]: I didn't get it. How does `definitions` parse a string into a value?

File: rfcs/90-parse-pairs-from-string.md (hunk 1 above)
Comment (xxchan) [reply rfcs/90-parse-pairs-from-string.md:35]: My fault. I didn't recall there's a mapping from string to type specified in specs

File: rfcs/90-parse-pairs-from-string.md (hunk 5 above)
Comment (xxchan) [reply rfcs/90-parse-pairs-from-string.md:9]: Why not? In my imagination, we support general parsing here and then config string follows.

File: rfcs/90-parse-pairs-from-string.md (hunk 3 above)
Comment (xxchan) [reply rfcs/90-parse-pairs-from-string.md:72]: > we only allow users to use these functions to do initialization

You mean only support something like `Open(config string) Storager`, instead of `Parse(config string) []Pair`? 
If so, users cannot add other pairs into config string, as you mentioned earlier.
If not, what's the difference? (btw I agree that parsing single pair API is too fine-grained)

File: rfcs/90-parse-pairs-from-string.md (hunk 2 above)
Comment (xxchan) [reply rfcs/90-parse-pairs-from-string.md:58]: Since from pair name and type in toml we can get a `WithXxx(v T)`, we should be able to get `T` from `string`. But how to parse `string` into the `v` is not decided, I think. (so we will need a `parseFunc`?)

Comment (xxchan) [reply rfcs/90-parse-pairs-from-string.md:58]: BTW, things like `typeMap` are private in `definitions`. How can we use them?

Comment (Xuanwo) [reply rfcs/90-parse-pairs-from-string.md:58]: > But how to parse string into the v is not decided, I think. 

It's what we need to specify in this GSP, for example:

From 

```
s3://<credential>@<endpoint>/workdir?key=value&default_storage_pairs.write=[key=value&key2=value]
```

To s3 and related pairs.

> BTW, things like typeMap are private in definitions.

We don't get them from `typeMap`.

Our code should be generated by `definitions` and we can get that information from specs and `service.toml`.

`service.toml` defined all pairs that this service supported.

File: rfcs/90-parse-pairs-from-string.md (hunk 3 above)
Comment (Xuanwo) [reply rfcs/90-parse-pairs-from-string.md:72]: For now, we have APIs like `s3.NewStorager(ps ..Pair) (Storager, error)` and we build a new API in service registry like `NewStorager(typ string, ps ...Pair) (Storager, error)`.

In this GSP, we will add functions that can: parse from string to type and pairs so that users can use them in this way:

```go
typ, ps, err := FromString("s3://xxxxx@yyyyy")
if err != nil {
    ...
}
store, err := NewStorager(typ, ps...)
```

> API should be defined in this GSP, and `FromString` is just an example.

And `FromString` is different from `FromMap` (please ignore the API I mentioned in above comment):

- `FromString` doesn't need an extra type field, we can add it into our format.
- `FromMap` needs an extra type field, like `FromMap(typ string, ps ...Pair)`

File: rfcs/90-parse-pairs-from-string.md (hunk 5 above)
Comment (Xuanwo) [reply rfcs/90-parse-pairs-from-string.md:9]: Let's build features that users need. For now, I don't think the user needs a pair parser except in service/storage initialization. Perhaps we can implement it when we find that there is a real need for it.

Comment (xxchan) [reply rfcs/90-parse-pairs-from-string.md:9]: I agree, but I thought if we provide API like `FromMap`, isn't it already a general pair parser? Users can use this API anywhere, although most often in initialization.

File: rfcs/90-parse-pairs-from-string.md (hunk 3 above)
Comment (xxchan) [reply rfcs/90-parse-pairs-from-string.md:72]: Then there's no misunderstanding. But this does not meet
> we only allow users to use these functions to do initialization.

File: rfcs/90-parse-pairs-from-string.md (hunk 2 above)
Comment (xxchan) [reply rfcs/90-parse-pairs-from-string.md:58]: Let me make things more clear.
In package `pairs`, we add suport to `parse(name string, value string) Pair`  (what puclic API is doesn't matter, internally there should be one like this), so we should 
- know `name` is a pair and its type `T`.  
- parse string into `T`

So in `pairs` we use a `PairMap` to register pairs. The specific registration code is generated by `definition`. `definition` knows everything and can generate type information both as string and as `T`, but what should be contained by `PairMap`? 

After code generation, `definitions` is useless. Then to parse string into `T` in `pairs`, we should know `T` and its parse function. If the registry `PairMap` stores string, how can we know what `T` is? And what's wrong that the parse function is `parseFunc func(string) (interface{}, error)`?

Comment (xxchan) [reply rfcs/90-parse-pairs-from-string.md:58]: TLDR: `definitions` only generate registration code which sets `PairMap`. Any other logic, include parsing `string` into `T` is written in `pairs`.

If this is not true, please tell me

Comment (xxchan) [reply rfcs/90-parse-pairs-from-string.md:58]: Oh I think we can
```go
typ := pairMap[name]
switch typ {
case "int": 
    pair.Value=parseInt(value)
case "bool":
    pair.Value=parseBool(value)
...
}
```
 
I used `reflect.Type` and `parseFunc func(string) (interface{}, error)` just to save the work here... I hoped to be more elegant and extensible, but actually this simple way is better, I think

Comment (Xuanwo) [reply rfcs/90-parse-pairs-from-string.md:58]: > Oh I think we can
> 
> ```go
> typ := pairMap[name]
> switch typ {
> case "int": 
>     pair.Value=parseInt(value)
> case "bool":
>     pair.Value=parseBool(value)
> ...
> }
> ```

Nice catch.

We don't need to support `parse(name string, value string) Pair`, we only need to parse the value and always transform them together instead of one by one.

File: rfcs/90-parse-pairs-from-string.md (hunk 5 above)
Comment (Xuanwo) [reply rfcs/90-parse-pairs-from-string.md:9]: It is indeed easy to misunderstand.

How about adding APIs like `NewStoragerFromString` which parse the string into pairs and call `NewStorager`? And we will not expose the API for parsing pairs until we figure this out.

Comment (xxchan) [reply rfcs/90-parse-pairs-from-string.md:9]: Then this API is effectively the old config string. We just polished the implementation.

Comment (Xuanwo) [reply rfcs/90-parse-pairs-from-string.md:9]: > Then this API is effectively the old init string. We just polished the implementation.

That's true.

Comment (Xuanwo) [reply rfcs/90-parse-pairs-from-string.md:9]: But the format could be redesigned.

File: rfcs/90-parse-pairs-from-string.md (hunk 2 above)
Comment (xxchan) [reply rfcs/90-parse-pairs-from-string.md:58]: OK, I'll update the proposal and demo later, although I'm still not 100% sure of what you mean. Anyway, let's delay the discussion after the update.

File: rfcs/90-parse-pairs-from-string.md (hunk 5 above)
Comment (xxchan) [reply rfcs/90-parse-pairs-from-string.md:9]: > But the format could be redesigned.

Oh actually this is what I wanted to consider later and that's why config string is not discussed in this proposal 🤔

**Comment by xxchan** [comment]
updated proposal and demo beyondstorage/go-storage#589 PTAL @Xuanwo 
(config string format is still not specified)

File: rfcs/90-support-initialization-via-config-string-and-map.md [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+But passing string config is indeed more convenient, if not much. And actually parsing string into pairs is not that difficult. We can use a registry mechanism like [GSP-48]'s implementation.
+
+## Proposal
+
+So I propose to support service init from config string and map:
+
+We add the following APIs:
+```go
+func NewServicerFromMap(ty string, m map[string]string) (types.Servicer, error) {}
```
Comment (Xuanwo) [inline rfcs/90-support-initialization-via-config-string-and-map.md:30]: How about split map support into another PR?

File: rfcs/90-support-initialization-via-config-string-and-map.md [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+`PairMap`s are generated.
+
+## Rationale
+
+Possible problems:
+
+1. In the pair registry `PairMap`, we may need more pair information beside types in the future, so we can also create a custom type `PairInfo` as the value type of the map. But to keep things simple, we'd better not overdesign now.
+
+2. It is possible that a service pair have the same name as a global pair. Internally, the name of a service pair has prefix `<type>_` so it wasn't a problem.
```
Comment (Xuanwo) [inline rfcs/90-support-initialization-via-config-string-and-map.md:85]: I come up with a new idea: we can remove the `<type>_` prefix from the pairs.

- we can check duplicate pairs via definitions.
- If the service pair conflicts with the global pair, the service pair will be a type alias to the global pair and marked as deprecated.
- In the service major version release, we can remove the service pair safely.

After these changes, we can remove the prefix safely.

File: rfcs/90-support-initialization-via-config-string-and-map.md [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+#### Pair Registry
+
+We register the types of global pairs and service pairs.
+
+From [go-storage] side:
+
+```go
+type PairMap map[string]string
+func RegisterGlobalPairMap(m PairMap) {}
```
Comment (Xuanwo) [inline rfcs/90-support-initialization-via-config-string-and-map.md:62]: After the service type prefix trimmed, maybe we don't need to `RegisterGlobalPairMap` and `RegisterServicePairMap` separately anymore?

File: rfcs/90-support-initialization-via-config-string-and-map.md [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+To support this, we have to know that a name is a pair (global or service), and its type, so we implement:
+
+#### Pair Registry
+
+We register the types of global pairs and service pairs.
+
+From [go-storage] side:
+
+```go
+type PairMap map[string]string
```
Comment (Xuanwo) [inline rfcs/90-support-initialization-via-config-string-and-map.md:61]: Maybe we don't need to export a new type here?

File: rfcs/90-support-initialization-via-config-string-and-map.md [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+func RegisterGlobalPairMap(m PairMap) {}
+func RegisterServicePairMap(ty string, m PairMap) {}
+```
+
+From services side, we can generate following code:
```
Comment (Xuanwo) [inline rfcs/90-support-initialization-via-config-string-and-map.md:66]: How about generating the map directly? like:

```
var pairMap = map[string]string{
      “xxxxx”: "xxxxx",
}
```

File: rfcs/90-support-initialization-via-config-string-and-map.md [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+#### Pair Registry
+
+We register the types of global pairs and service pairs.
+
+From [go-storage] side:
+
+```go
+type PairMap map[string]string
+func RegisterGlobalPairMap(m PairMap) {}
+func RegisterServicePairMap(ty string, m PairMap) {}
```
Comment (Xuanwo) [inline rfcs/90-support-initialization-via-config-string-and-map.md:63]: I prefer `RegisterServiceSchema` here.

File: rfcs/90-support-initialization-via-config-string-and-map.md (hunk 6 above)
Comment (xxchan) [reply rfcs/90-support-initialization-via-config-string-and-map.md:30]: Then I prefer to support map first since I plan to use it to support config string...

File: rfcs/90-support-initialization-via-config-string-and-map.md (hunk 9 above)
Comment (xxchan) [reply rfcs/90-support-initialization-via-config-string-and-map.md:61]: Fine

File: rfcs/90-support-initialization-via-config-string-and-map.md (hunk 6 above)
Comment (Xuanwo) [reply rfcs/90-support-initialization-via-config-string-and-map.md:30]: After discussion at [#specs-gsp-90:aos.dev](https://matrix.to/#/#specs-gsp-90:aos.dev), current design is OK to me.

How about just export `NewServicerFromString` to users and don't implement `NewServicerFromMap` for now? (use map internally is LGTM)

File: rfcs/90-support-initialization-via-config-string-and-map.md (hunk 7 above)
Comment (xxchan) [reply rfcs/90-support-initialization-via-config-string-and-map.md:85]: I found that actually, we cannot add a conflicting pair in `service.toml` now. (generate will fail with "pair conflict").

Then
- What was the use of `<type>_`?
- We won't have pair aliases as you mentioned here

Comment (Xuanwo) [reply rfcs/90-support-initialization-via-config-string-and-map.md:85]: > What was the use of <type>_?

`<type>_` is used to avoid pairs overwrite, but we can fix it by checking them while generation.

> We won't have pair aliases as you mentioned here

Yes, there are not implemented for now.

We can generate code for conflict pairs like:

```go
var WithStorageClass = pairs.WithStorageClass
```

Comment (xxchan) [reply rfcs/90-support-initialization-via-config-string-and-map.md:85]: You mean we will add support for conflicting pairs? Why?

Comment (xxchan) [reply rfcs/90-support-initialization-via-config-string-and-map.md:85]: Maybe I need to investigate the generation logic carefully first

Comment (Xuanwo) [reply rfcs/90-support-initialization-via-config-string-and-map.md:85]: > You mean we will add support for conflicting pairs? Why?

There is no need for now, but in the future, some service pairs could become global pairs, and we need to have a migration plan.

File: rfcs/90-support-initialization-via-config-string-and-map.md (hunk 8 above)
Comment (xxchan) [reply rfcs/90-support-initialization-via-config-string-and-map.md:62]: What if service A has pair `(P, T1)`, but service B has pair `(P, T2)`, and `T1 != T2`?

If you mean all the pairs will share the same namespace after prefix trimmed, I think this may be dangerous.

Comment (xxchan) [reply rfcs/90-support-initialization-via-config-string-and-map.md:62]: Oh, this can be checked in `init` here

Comment (Xuanwo) [reply rfcs/90-support-initialization-via-config-string-and-map.md:62]: I don't think they are in the same namespace:

- If `P` is a global pair, `T1 == T2` is ensured.
- If `P` is a service pair, user must use them in `A.WihtP` and `B.WithP`, and I think it's OK to allow `T1 != T2`?

Comment (xxchan) [reply rfcs/90-support-initialization-via-config-string-and-map.md:62]: If they are not in the same namespace, we will still need `RegisterServicePairMap`

Comment (Xuanwo) [reply rfcs/90-support-initialization-via-config-string-and-map.md:62]: Well, we do need `RegisterServicePairMap`, but we can register global pairs together instead of another public API `RegisterGlobalPairMap`.

Comment (xxchan) [reply rfcs/90-support-initialization-via-config-string-and-map.md:62]: Oh, you mean global pairs are included in any service pair map

Comment (Xuanwo) [reply rfcs/90-support-initialization-via-config-string-and-map.md:62]: Yes, global pairs are shared across all services and they can be included in any service pair map. (We can check them while doing generation)

**Comment by Xuanwo** [comment]
`DefaultPairs` and `Features` could be complex to parse and define, It's OK to me to split them into another GSP.

**Comment by xxchan** [comment]
draft updated, details and tests to be polished
PTAL at the design and the parser (https://github.com/beyondstorage/go-storage/pull/589) first @Xuanwo

File: rfcs/90-re-support-initialization-via-config-string.md [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+The format of the config string is (optional parts marked by squared brackets):
+
+`<type>://[<name>][<work_dir>][?key1=value1&...&keyN=valueN]`
+
+- name: storage name, e.g., bucket name. MUST NOT contain /
+- work_dir: For object storage, it is prefix; for fs, it is directory path. MUST start with / for every storage services.
+
+So a valid config string could be:
+
+- `qingstor://bucket_name`
```
Comment (Xuanwo) [inline rfcs/90-re-support-initialization-via-config-string.md:45]: Let's use `s3` / `gcs` for examples.

File: rfcs/90-re-support-initialization-via-config-string.md [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+To support this, we have to know that a name is a pair (global or service), and its type, so we implement:
+
+#### Pair Registry
+
+We register the types of global pairs and service pairs.
+
+From [go-storage] side:
+
+```go
+func RegisterServiceSchema(ty string, m map[string]string) {}
```
Comment (Xuanwo) [inline rfcs/90-re-support-initialization-via-config-string.md:81]: Let's mark it clear that this function should never be called by users.

File: rfcs/90-re-support-initialization-via-config-string.md [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+From services side, we can generate following code:
+
+```go
+var pairMap = map[string]string{
+	"xxxxx": "xxxxx",
+}
+
+func init() {
+	// ...
+	pairs.RegisterServiceSchema("<type>", pairMap)
```
Comment (Xuanwo) [inline rfcs/90-re-support-initialization-via-config-string.md:93]: How about put this function in `services` like `services.RegisterSchema`?

**Comment by Xuanwo** [comment]
Nice job!

File: rfcs/90-re-support-initialization-via-config-string.md [hunk 15]
```diff
@@ -0,0 +1,139 @@
+---
+author: xxchan <xxchan22f@gmail.com>
+status: draft
+updated_at: 2021-06-09
+deprecates:
+  - 13-remove-config-string.md
+---
+
+# GSP-90: Re-support Initialization Via Config String
```
Comment (Xuanwo) [inline rfcs/90-re-support-initialization-via-config-string.md:9]: How about naming `Initialization Via Connection String`?

`connection string` looks widely used in computer science: https://en.wikipedia.org/wiki/Connection_string

Comment (xxchan) [reply rfcs/90-re-support-initialization-via-config-string.md:9]: I don't remember why I name it `config string` 🤔

Comment (xxchan) [reply rfcs/90-re-support-initialization-via-config-string.md:9]: Oh, it was the old name

File: rfcs/90-re-support-initialization-via-config-string.md (hunk 14 above)
Comment (xxchan) [reply rfcs/90-re-support-initialization-via-config-string.md:93]: Oh, actually it is in `services`. not updated here

File: rfcs/90-re-support-initialization-via-config-string.md (hunk 13 above)
Comment (xxchan) [reply rfcs/90-re-support-initialization-via-config-string.md:81]: I'm wondering is it possible to enforce this requirement 🤔

Comment (Xuanwo) [reply rfcs/90-re-support-initialization-via-config-string.md:81]: I think it's impossible.

File: rfcs/90-re-support-initialization-via-connection-string.md [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+- name: storage name, e.g., bucket name. MUST NOT contain /
+- work_dir: For object storage, it is prefix; for fs, it is directory path. MUST start with / for every storage services.
+- At least one of name and work_dir MUST be present.
+- For the `key=value` pairs, if `=value` is missing, we just ignore the pair. But `key=` means a pair with a blank value.
+
+So a valid connection string could be:
+
+- `s3://bucket_name`
+- `s3://bucket_name/prefix`
+- `s3://bucket_name/prefix?credential=hmac:xxxx:xxxx&endpoint=http://bucket_name.s3-website-Region.amazonaws.com`
```
Comment (Xuanwo) [inline rfcs/90-re-support-initialization-via-connection-string.md:49]: Maybe s3 endpoint should not contain `bucket_name`? As described in [Amazon Simple Storage Service endpoints and quotas](https://docs.aws.amazon.com/general/latest/gr/s3.html):

```
s3.us-east-1.amazonaws.com
s3-fips.us-east-1.amazonaws.com
s3.amazonaws.com
s3.dualstack.us-east-1.amazonaws.com**
```

File: rfcs/90-re-support-initialization-via-connection-string.md [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+
+We register the types of global pairs and service pairs.
+
+From [go-storage] side:
+
+```go
+// RegisterServiceSchema will register a service's pair map.
+//
+// Users SHOULD NOT call this function.
+func RegisterServiceSchema(ty string, m map[string]string) {}
```
Comment (Xuanwo) [inline rfcs/90-re-support-initialization-via-connection-string.md:86]: So we only support init a `Storager` from the connection string? Do we need to support init `Servicer`? `RegisterServiceSchema` may be confused between `servicer` and `storager`?

> BTW, It's OK to support storage only in this GSP if we have left some space for service support.

File: rfcs/90-re-support-initialization-via-connection-string.md [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+}
+```
+
+## Rationale
+
+### Alternative Format
+
+We can also add `<credential>@<endpoint>` in the config string, i.e.:
+
+`type://<credential>@<endpoint>/<name><work_dir>?key1=value1&...&keyN=valueN`
```
Comment (Xuanwo) [inline rfcs/90-re-support-initialization-via-connection-string.md:108]: I prefer this style, but you are the boss. Let's discuss this in another GSP.

File: rfcs/90-re-support-initialization-via-connection-string.md [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+
+However, our `credential` and `endpoint` is different from the `username:password@hostname` in URL, and we cannot easily utilize `url.Parse` to reduce our work. So we just keep the format simple currently, and we can add them in the future if really needed.
+
+### Richer Schema?
+
+In the pair registry, we may need more pair information beside types in the future, so we can also create a custom type `PairInfo` as the value type of the map. But to keep things simple, we'd better not overdesign now.
```
Comment (Xuanwo) [inline rfcs/90-re-support-initialization-via-connection-string.md:114]: LGTM

File: rfcs/90-re-support-initialization-via-connection-string.md (hunk 17 above)
Comment (xxchan) [reply rfcs/90-re-support-initialization-via-connection-string.md:86]: There's no difference between `servicer` and `storager`.

I proviced `NewServicerFromString` and `NewStoragerFromString`. They both call `parseString(connStr)` which relys on `servicePairMaps`, which is initialized by `RegisterServiceSchema`

Comment (Xuanwo) [reply rfcs/90-re-support-initialization-via-connection-string.md:86]: Well, I got it.

How about rename `services.RegisterServiceSchema` to `services.RegisterSchema` so that we can avoid this confusion?

Comment (Xuanwo) [reply rfcs/90-re-support-initialization-via-connection-string.md:86]: Something missing: the format could be different for service and storage.

`s3` service doesn't need `name` nor `work_dir`, the examples may be:

```
`s3://?credential=hmac:xxxx:xxxx&endpoint=http://s3.us-east-2.amazonaws.com`
```

Comment (xxchan) [reply rfcs/90-re-support-initialization-via-connection-string.md:86]: > the format could be different for service and storage.

Oops. Then we just allow both of them to be missing? Let this be checked in generated service code, like other `required` pairs.

Comment (Xuanwo) [reply rfcs/90-re-support-initialization-via-connection-string.md:86]: I think only `name` is required for `Storage`, if the name exists, it's a storage, or it's a service. And `work_dir` has a default value `/`.

For our formats, It's OK to add omit the workdir like:

- `s3://bucketname?credential=hmac:xxxx:xxxx&endpoint=http://s3.us-east-2.amazonaws.com`
- `fs://` & `fs:///`

Allow omit or not are both OK to me, I respect your decision.

File: rfcs/90-re-support-initialization-via-connection-string.md [hunk 20]
```diff
@@ -0,0 +1,154 @@
+---
+author: xxchan <xxchan22f@gmail.com>
+status: draft
+updated_at: 2021-06-09
```
Comment (Xuanwo) [inline rfcs/90-re-support-initialization-via-connection-string.md:4]: Update the `updated_at`

**Review by Xuanwo — APPROVED** [review]

File: rfcs/90-re-support-initialization-via-connection-string.md [hunk 21]
```diff
@@ -0,0 +1,154 @@
...
+### Format
+
+The format of the connection string is (optional parts marked by squared brackets):
+
+`<type>://[<name>][<work_dir>][?key1=value1&...&keyN=valueN]`
+
+- name: storage name, e.g., bucket name. MUST NOT contain /
+- work_dir: For object storage, it is prefix; for fs, it is directory path. MUST start with / for every storage services.
+- For the `key=value` pairs, if `=value` is missing, we just ignore the pair. But `key=` means a pair with a blank value.
+- If their are multiple pairs with the same key, the first one will be picked.
```
Comment (JinnyYi) [inline rfcs/90-re-support-initialization-via-connection-string.md:43]: Is it necessary to give the naming conventions for key? Is the key case sensitive or must match the definitions exactly?

Comment (xxchan) [reply rfcs/90-re-support-initialization-via-connection-string.md:43]: In current implementation, it is the same as the name in `toml` file

Comment (xxchan) [reply rfcs/90-re-support-initialization-via-connection-string.md:43]: I can specify this in the proposal.

**Review by JinnyYi — APPROVED** [review]
